Token class updates to accommodate new layout features

bubblobill opened the pull request:
resolves #5103 
Progresses #5096 
Progresses #3691
Includes most of #5097 

### Description of the Change
New property
New property macro functions
Refactoring

### Possible Drawbacks
none

### Documentation Notes
n/a

### Release Notes
n/a

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/RPTools/maptool/5104)
<!-- Reviewable:end -->

bubblobill committed:
Token class:
clean up some property names, e.g. isFlippedIso to flippedIso so that getter becomes isFlippedIso().

Unnecessary Token removed from property tokenOpacity. Now just opacity.

Reordered to keep layout props together.
Added new layout prop imageRotation

Updated tokenLayoutProp macro functions

bubblobill committed:
Merge remote-tracking branch 'origin/TokenLayout' into TokenLayout

bubblobill committed:
Updated TokenDto to handle refactored property names and new property imageRotation

bubblobill committed:
Updated Token again due to mysterious update overwriting it

bubblobill committed:
Many things coming together

bubblobill committed:
Merge remote-tracking branch 'origin/TokenLayout' into TokenLayout

bubblobill commented:
Yay, it runs, and I am reminded why methods should be properly named the first time.
Now I just have to rewrite ZoneRenderer

bubblobill committed:
Forgot I hadn't finished hexes

cwisniew commented on src/main/java/net/rptools/maptool/client/ui/token/dialog/edit/EditTokenDialog.java:
```diff
@@ -940,10 +968,10 @@ public boolean commit() {
+    //    MapTool.getFrame()
+    //        .getCurrentZoneRenderer()
+    //        .getZone()
+    //        .tokenMaskTopologyChanged(token.getMaskTopologyTypes());
```
Please remove rather than just commenting out :)

bubblobill replied:
Points at` // private CharSheetController controller;`

cwisniew replied:
You can remove that one while you are there if you like :) feel free to point out others you find :)

bubblobill replied:
Think I caught them all

cwisniew commented on src/main/java/net/rptools/maptool/client/ui/token/dialog/edit/TokenLayoutPanelHelper.java:
```diff
@@ -0,0 +1,1317 @@
+  // @formatter:off
+  // spotless:off
+    // Component Getters
+    public JComboBox getSizeCombo(){ if (sizeCombo == null) sizeCombo = (JComboBox) parent.getComponent("size"); return sizeCombo; }
```
just format these as expected rather than turning off spotless :)

bubblobill replied:
Not until it gets out of draft status

bubblobill replied:
Done. Code uglified.

cwisniew commented on src/main/java/net/rptools/maptool/client/ui/token/dialog/edit/TokenLayoutPanelHelper.java:
```diff
@@ -0,0 +1,1317 @@
+    getZoomSpinner().setValue(1d);
+  }
+
+  String scaleStartSVG =
+      "<svg version=\"1.1\" height=\"32\" width=\"32\" viewBox=\"0 0 32 32\" xmlns=\"http://www.w3.org/2000/svg\"><g stroke=\"black\">";
+  String scaleNSSVG =
+      "<path id=\"NS\" d=\"m15 5.5-5 1.6 6-6.1 6 6.1-5-1.6v21l5-1.7-6 6.1-6-6.1 5 1.7z\" fill=\"black\" stroke-width=\"1\"/>";
+  String scaleEWSVG =
+      "<path id=\"EW\" d=\"m27 15-1.6-5 6.1 6-6.1 6 1.6-5h-21l1.7 5-6.1-6 6.1-6-1.7 5z\" fill=\"black\" stroke-width=\"1\"/>";
+  String scaleEndSVG = "<circle id=\"hub\" cx=\"16\" cy=\"16\" r=\"4\" fill=\"white\" /></g></svg>";
+
+  public static ImageIcon makeSVGIcon(int size, String svg) {
+    InputStream stream = new ByteArrayInputStream(svg.getBytes(StandardCharsets.UTF_8));
+    FlatSVGIcon.ColorFilter colorFilter = new FlatSVGIcon.ColorFilter();
+    // Convert a the original color into another one supposedly
+    colorFilter.add(Color.BLACK, UI_DEFAULTS.getColor(FlatIconColors.OBJECTS_BLACK_TEXT));
+    colorFilter.add(Color.WHITE, UI_DEFAULTS.getColor(FlatIconColors.ACTIONS_BLUE_DARK));
+    try {
+      FlatSVGIcon flatSVGIcon = new FlatSVGIcon(stream);
+      flatSVGIcon.setColorFilter(colorFilter);
+      return flatSVGIcon.derive(size, size);
+    } catch (IOException e) {
+      return null;
+    }
+  }
+
+  ImageIcon[] scaleIcons = new ImageIcon[3];
+
+  private void createButtonIcons() {
+    scaleIcons[0] = makeSVGIcon(ICON_SIZE, scaleStartSVG + scaleNSSVG + scaleEWSVG + scaleEndSVG);
+    scaleIcons[1] = makeSVGIcon(ICON_SIZE, scaleStartSVG + scaleEWSVG + scaleEndSVG);
+    scaleIcons[2] = makeSVGIcon(ICON_SIZE, scaleStartSVG + scaleNSSVG + scaleEndSVG);
+  }
```
Its better to add these as resource files and add the functions to load them ResourceManager class, let me know if you need a hand doing this

bubblobill replied:
Inclined to disagree. As you can see in these images. Dark themes do not change the icon colours like they are supposed to.
![image](https://github.com/user-attachments/assets/9c453709-28f1-4e8f-8381-08b7f36082f9)
![image](https://github.com/user-attachments/assets/da032b7b-38d1-4098-aa11-56593b5ef6eb)
The FlatLaF docs are very poor when it comes to creating responsive SVG. So unless I can get it to work I will probably end up hard-coding the colours into the SVG at run-time.

cwisniew replied:
https://github.com/JFormDesigner/FlatLaf/issues/402
You have to make sure you are using one of the defined colours, like the current icons are.

No embedding SVG strings into the code please it makes it too hard to edit them

bubblobill replied:
Icons added to resources

bubblobill committed:
Changed all comments to block comments to make it easier to find commented code.
Commented code removed

bubblobill committed:
Corrected squares in createGridShape.

bubblobill committed:
Added image rotation property to layoutProps functions

bubblobill committed:
Sanity saving string concatenation

bubblobill committed:
Spotless indent

bubblobill committed:
Code cleanup

bubblobill committed:
Code cleanup

bubblobill committed:
Refactoring

bubblobill committed:
Refactoring

bubblobill committed:
Refactoring

bubblobill committed:
Refactoring

bubblobill committed:
Refactoring
Token drawing moved to TokenRenderer
Facing-arrow painting moved to FacingArrowRenderer
Some functionality moved to ImageUtils

bubblobill committed:
Facing-arrow painting moved to FacingArrowRenderer

bubblobill committed:
Code cleanup
Remove old layout panel

bubblobill commented:
In order to get tokens to paint properly I had to mess with ZoneRenderer. The renderTokens() code in ZR is a nightmare of repetitious code and lots of fragile non-optimal stuff.

Since ZR needs major splitting anyway, I created TokenRenderer. At present it only deals with painting tokens and I left all the weird stuff I don't understand properly behind.
TokenRenderer checks if there have been changes to the token that would impact the rendered image (layout props) and only creates new images when required. 

No idea if it is faster, but image operations are performed in the following sequence with the intention of maintaining the best quality; Flip X/Y -> Flip Iso -> Scale[X, Y, XY, Footprint, and Zoom (all at once to prevent multiple scaling lossiness)] -> Rotate.

Image caches are cleared on change of zoom. Not sure if it is necessary but it seemed safer.

There are also problems with token bounds and locations. The whole everything works this way except for squares annoys the crap out of me as Everything I do is based on the token's actual position. Images are transformed such that the origin lies at their centre +/- any anchor. Translate the graphics object to the token position, scale and rotate as required and everything is easy.
With bounds, tokens have a bounding area defined by their footprint AND a bounding area that covers the token image. Think of a tree whose picture includes its canopy but the space occupied by the trunk is its footprint. Currently getBounds on a token returns the size of the whole image including offsets and scaling.
I seriously wanted to change this because it results in other things being drawn incorrectly using StG.

This is what happens if halos are drawn on offset images. Both tokens are StG. This is what the shark should look like.
![image](https://github.com/user-attachments/assets/9e01df7f-8c87-4049-8311-7516b7ec5dda)
But it snaps to the wrong place.
![image](https://github.com/user-attachments/assets/17018599-98fc-4ddf-aa9a-b208bfa5a500)

Since it has always been the case I didn't mess with it. But if all this gets through then I am mightily inclined to change it.

bubblobill committed:
Code cleanup
Icons embedded

bubblobill committed:
Spotless exception removed

bubblobill committed:
Merge branch 'develop' into TokenLayout

bubblobill committed:
Merge branch 'develop' into TokenLayout

bubblobill committed:
Fixed getScaledTokenImage returning incorrect scaling for free-size images

bubblobill committed:
cleanup and spotless

bubblobill committed:
cleanup and spotless

bubblobill committed:
Merge branch 'develop' into TokenLayout

bubblobill committed:
Merge branch 'RPTools:develop' into TokenLayout

bubblobill committed:
Merge branch 'RPTools:develop' into TokenLayout

kwvanderlinde commented on src/main/java/net/rptools/lib/MathUtil.java:
```diff
@@ -0,0 +1,150 @@
+import org.apache.logging.log4j.Logger;
+
+/* Utility class for useful mathematical methods */
+public class MathUtil {
```
We can do without most of this class as the methods are either unused or can be simplified. I'll leave specific comments below.

bubblobill replied:
Agreed. I originally created it to house mathematical methods that I found lurking in other classes that I could never find when I wanted them. Then I gave up trying to find existing methods and just started adding methods as they became useful. Some of which I stopped using as my code developed, as you discovered.

kwvanderlinde commented on src/main/java/net/rptools/lib/MathUtil.java:
```diff
@@ -0,0 +1,150 @@
+  private static final Logger log = LogManager.getLogger(MathUtil.class);
+
+  /**
+   * Faster version of absolute for integers
```
Any proof of this claim? `Math.abs()` calls are always inlined with platform-specific instructions, so there's no branching concerns with it.

Anyways, this method isn't used and can be removed.

bubblobill replied:
Yes there is evidence, probably buried in StackExchange. Although it is probably easier to just use FastMath, and as you noticed, my need for it disappeared.

kwvanderlinde commented on src/main/java/net/rptools/lib/MathUtil.java:
```diff
@@ -0,0 +1,150 @@
+  }
+
+  /**
+   * Faster version of absolute
```
I guarantee this is slower than `Math.abs()`. Plus this method isn't used.

kwvanderlinde commented on src/main/java/net/rptools/lib/MathUtil.java:
```diff
@@ -0,0 +1,150 @@
+   * @return the equivalent value of valueToMap in the target range
+   * @param <T>
+   */
+  public static <T extends Number> T mapToRange(
```
The generics don't really gain us anything here since internally it has to check `float` vs `double` anyways. I.e., an overload would do just as well.

But it can also just be replaced with a purely `double` version. The one caller that uses `float` can cast the result.

bubblobill replied:
\<whine>But I don't like casting and like generics... \</whine>

kwvanderlinde commented on src/main/java/net/rptools/lib/MathUtil.java:
```diff
@@ -0,0 +1,150 @@
+              ((mapValue.doubleValue() - inMin.doubleValue())
+                      * (outMax.doubleValue() - outMin.doubleValue())
+                      / (inMax.doubleValue() - inMin.doubleValue())
+                  + outMin.doubleValue());
```
This kind of lerp implementation tends to be inaccurate when `valueToMap == in_max`, due to floating point shenanigans. An alternative with accurate endpoints would be:
```java
var t = (valueToMap - in_min) / (in_max - in_min);
result = (1 - t) * out_min + t * out_max;
```

bubblobill replied:
good thinking 99

kwvanderlinde commented on src/main/java/net/rptools/lib/MathUtil.java:
```diff
@@ -0,0 +1,150 @@
+   * @param highBound
+   * @return
+   */
+  public static int constrainInt(int value, int lowBound, int highBound) {
```
This method is the same as `Math.clamp(long, int, int)` and can be removed in favour of that.

bubblobill replied:
wilco

kwvanderlinde commented on src/main/java/net/rptools/lib/MathUtil.java:
```diff
@@ -0,0 +1,150 @@
+   * @return
+   * @param <T>
+   */
+  public static <T extends Number> T constrainNumber(T value, T lowBound, T highBound) {
```
This method is equivalent to `Math.clamp(double, double, double)` and can be removed.

bubblobill replied:
wilco

kwvanderlinde commented on src/main/java/net/rptools/lib/MathUtil.java:
```diff
@@ -0,0 +1,150 @@
+    return d;
+  }
+
+  public static boolean isDouble(Object o) {
```
With the other proposed simplifications, `isDouble()`, `isFloat()` and `isNumber()` are not needed.

bubblobill replied:
spoilsport

kwvanderlinde commented on src/main/java/net/rptools/lib/MathUtil.java:
```diff
@@ -0,0 +1,150 @@
+   * @return truncated double value
+   */
+  public static double doublePrecision(double value, int decimalPlaces) {
+    double d = Double.parseDouble(String.format("%." + decimalPlaces + "f", value));
```
This is fine for the UI usage, but outside of that will be a pretty slow implementation. Can use this instead:
```
double divisor = Math.pow(10, -decimalPlaces);
double d = divisor * Math.round(value / divisor);
```

bubblobill replied:
Where were you when I was researching for a fast and reliable Math.trunc?

kwvanderlinde commented:
I took this for a test and found a bug regarding halos. A token with no halo is shown as having the "white" halo when the menu is opened.
![white-halo](https://github.com/user-attachments/assets/e1542501-e3ca-4274-b75c-b2d97176c499)

kwvanderlinde commented on src/main/java/net/rptools/maptool/model/Token.java:
```diff
@@ -292,16 +287,15 @@ public enum Update {
 
   // Jamz: allow token alpha channel modification
-  private @Nonnull Float tokenOpacity = 1.0f;
+  private @Nonnull Float opacity = 1.0f;
```
Can't rename fields or token settings in existing campaigns will be lost (applies to `opacity`, `isFlippedX`, `isFlippedY`, `isFlippedIso`).

It's fine that the accessor methods or DTO fields are renamed, just the model fields need to remain the same.

bubblobill replied:
yeah... found that out the hard way.

kwvanderlinde commented:
Got partyway through reviewing. Will pick it up again later today.

The feature is looking quite good. The new layout functionality feels very nice to use.

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/functions/TokenPropertyFunctions.java:
```diff
@@ -929,24 +930,85 @@ public Object childEvaluate(
+      String rotation = String.valueOf(token.getImageRotation());
+      String scaleX = String.valueOf(token.getScaleX());
+      String scaleY = String.valueOf(token.getScaleY());
+      String footprintScaleValue =
```
These should all be left as numeric values. Otherwise the JSON will contain a bunch of strings that will then require further parsing by the caller.

bubblobill replied:
roger

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/functions/TokenPropertyFunctions.java:
```diff
@@ -929,24 +930,85 @@ public Object childEvaluate(
+        sb.append("scaleX=" + scaleX + delim);
+        sb.append("scaleY=" + scaleY + delim);
+        sb.append("footprintScale=" + footprintScaleValue);
+        return sb.toString();
```
Since you're using a `StringBuilder` anyways, might as well avoid the (expensive) string concats and chain the appends. E.g.,
```java
sb.append("scale=").append(scale).append(delim);
```

bubblobill replied:
roger

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/functions/TokenPropertyFunctions.java:
```diff
@@ -929,24 +930,85 @@ public Object childEvaluate(
+    /*
+     * setExtendedTokenLayoutProps(StrProp/JSON Object, token: currentToken(), mapName = current map)
+     */
+    if (functionName.equalsIgnoreCase("setExtendedTokenLayoutProps")) {
```
Anyway this could be folded into `setTokenLayoutProps()`? The set of options parallels the output of `getTokenLayoutProps()` and so would be easier for macro writers to discover.

bubblobill replied:
mebbe. it's been so long I will have to revisit and work out why I did it this way.

bubblobill replied:
Had a look to find out why I made it separate. 
`setTokenLayoutProps()` has a fixed arg list `setTokenLayoutProps(scale, xOffset, yOffset, token, map)`.
dealing with a bunch of optional arguments would have been messy `(scale, xOffset, yOffset, scaleX, scaleY, rotation, token, map)` and made it even harder to add future changes I have considered like skew and perspective adjustmenets.

Thus I opted to replace it with a function that performs the same actions but can accommodate the new and possible future changes by simply taking `(props_map, token, map)`.

The name needs help, I think I will rename it to `setLayoutProps`

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/ui/zone/renderer/tokenRender/TokenRenderer.java:
I'm not sure what changed, but I'm finding that "Flip > Isometric Plane" doesn't result in correct rendering. The token is typically rendered flat, but once it starts being dragged it becomes iso.
![flip-iso](https://github.com/user-attachments/assets/0684512d-91ff-478e-a125-ee869f7894b5)

bubblobill replied:
Probably ignoring a condition somewhere like token shape

bubblobill replied:
This is part of the existing code net.rptools.maptool.client.ui.zone.renderer.ZoneRenderer#showBlockedMoves that does not yet use TokenRenderer, i.e. it wasn't me.

kwvanderlinde commented on src/main/java/net/rptools/lib/image/ImageUtil.java:
```diff
@@ -47,64 +42,135 @@
+        try {
+          return ImageUtil.scaleBufferedImage(img, outputWidth, outputHeight);
+        } catch (Exception e) {
+          log.debug(e.getLocalizedMessage(), e);
```
Is there an actual error that we should be concerned about here? At a glance, I don't think the `ResampleOp` will throw anything in our case.

If there is a worry, this should be a `log.error()`, or at least a `log.warn()`.

bubblobill replied:
IIRC, it was bugging out on images with extreme aspect ratios

kwvanderlinde commented on src/main/java/net/rptools/lib/image/ImageUtil.java:
```diff
@@ -47,64 +42,135 @@
+        return ImageUtil.scaleBufferedImage(
+            img, (int) Math.ceil(b.width * zoom), (int) Math.ceil(b.height * zoom));
+      } catch (Exception e) {
+        log.debug(e.getLocalizedMessage(), e);
```
Same as above, either don't handle the error or make it `log.error()`.

bubblobill replied:
you and your semantic correctness

kwvanderlinde commented on src/main/java/net/rptools/lib/image/ImageUtil.java:
```diff
@@ -389,20 +407,41 @@ public static BufferedImage createOutline(BufferedImage sourceImage, Color color
+   * @param token The token containing the flip-states
+   * @return A modified image, or the original image if no processing performed.
+   */
+  public static BufferedImage flipTokenImage(BufferedImage image, Token token) {
```
This method isn't used.

bubblobill replied:
disappointing after all the time I spend writing methods that I find they already exist somewhere else.

kwvanderlinde commented on src/main/java/net/rptools/lib/image/ImageUtil.java:
```diff
@@ -415,21 +454,203 @@ public static BufferedImage flip(BufferedImage image, int direction) {
+    return bytesToImage(dataStream.toByteArray(), image);
+  }
+
+  public static double getIsoFigureHeightOffset(Token token, Rectangle2D footprintBounds) {
```
`getIsoFigureHeightOffset()` and `getIsoFigureScaleFactor()` seem like they would be more natural in the `Token` class as they don't really have to do with image manipulation even though they are fed into token image rendering.

bubblobill replied:
Their only value is for working out where to draw the image and were extracted from the guts of zoneRenderer. Personally I think Token is already bloated and want it to only have the x/y of the image centre/footprint origin and separate all the bits for rendering. They probably belong in Grid or Footprint. I'll look into their utility before considering a relocation.

bubblobill replied:
They are used in separate places which is why I brought them together into the utility class.

bubblobill committed:
ImageUtil:
- removed unused method.
- changed log messages to warn
MathUtil:
- Removed generics
- Removed unused and duplicate functions
Token:
- reverted renamed properties

bubblobill commented:
Implemented the majority of requested changes, but too much time has passed and now there are conflicts I don't know how to resolve.

kwvanderlinde commented:
> Implemented the majority of requested changes, but too much time has passed and now there are conflicts I don't know how to resolve.

I can help you out with this. Does your `TokenLayout` branch have all your latest changes? If so, I can pull it, merge with `develop` and resolve these conflicts.

bubblobill committed:
ImageUtil:
- removed unused method.
- changed log messages to warn
MathUtil:
- Removed generics
- Removed unused and duplicate functions
Token:
- reverted renamed properties
Changed MT Function to setLayoutProps

bubblobill commented:
> > Implemented the majority of requested changes, but too much time has passed and now there are conflicts I don't know how to resolve.
> 
> I can help you out with this. Does your `TokenLayout` branch have all your latest changes? If so, I can pull it, merge with `develop` and resolve these conflicts.

Implementing the changes broke the layout panel. The pairing of spinners and sliders connected to wheel events no longer works, and I now have disproportionate movement when dragging the token image.

Methinks your requested changes are questionable :)

bubblobill committed:
Fixed mapToRange function
Stopped image drag mouse movement rounding to zero
Reworked SpinnerSliderPaired
Removed most uses of MathUtil.doublePrecision because using numbers
instead of string format produces inequitable values.
Fixed rotation guide not showing for negative values.

bubblobill commented:
> > Implemented the majority of requested changes, but too much time has passed and now there are conflicts I don't know how to resolve.
> 
> I can help you out with this. Does your `TokenLayout` branch have all your latest changes? If so, I can pull it, merge with `develop` and resolve these conflicts.

Think I have fixed everything. TokenLayout branch now has a (theoretical) publishable version.

kwvanderlinde committed:
Merge branch 'develop' into TokenLayout

kwvanderlinde commented:
> > > Implemented the majority of requested changes, but too much time has passed and now there are conflicts I don't know how to resolve.
> > 
> > 
> > I can help you out with this. Does your `TokenLayout` branch have all your latest changes? If so, I can pull it, merge with `develop` and resolve these conflicts.
> 
> Think I have fixed everything. TokenLayout branch now has a (theoretical) publishable version.

I've resolved the merge conflicts. Had to do it with the GitHub editor though, so didn't have access to spotless. @bubblobill Mind running `spotlessApply`?

In the meantime I'll review the updates.

bubblobill committed:
Spotless

bubblobill commented:
> I've resolved the merge conflicts. Had to do it with the GitHub editor though, so didn't have access to spotless. @bubblobill Mind running `spotlessApply`?

done

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/functions/TokenPropertyFunctions.java:
```diff
@@ -113,7 +113,8 @@ private TokenPropertyFunctions() {
         "getTokenLayoutProps",
-        "setTokenLayoutProps",
+        "setLayoutProps",
+        "setExtendedTokenLayoutProps",
```
We'll need `setTokenLayoutProps` to remain for backwards compatibility.

I assume you meant to rename `setExtendedTokenLayoutProps` to `setLayoutProps` instead?

bubblobill replied:
> I assume you meant to rename setExtendedTokenLayoutProps to setLayoutProps instead?

Correct. It seems one cannot truly trust find/replace

bubblobill replied:
https://github.com/RPTools/maptool/pull/5104/commits/17efdde4e4b48fae17d02489568974fbd66bb254

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/ui/token/dialog/edit/TokenLayoutPanelHelper.java:
```diff
@@ -0,0 +1,1341 @@
+        evt -> {
+          log.info(evt);
+          if (evt.getPropertyName().toLowerCase().contains("value")
+              && evt.getNewValue().getClass().isAssignableFrom(Double.class)
```
This `isAssignableFrom()` check is backwards. But it can also be cleared up by replacement with `instanceof`:
```
          if (evt.getPropertyName().toLowerCase().contains("value")
              && evt.getNewValue() instanceof Number newValue
              && (newValue.doubleValue() < 0.34 || newValue.doubleValue() > 3d)) {
            ...
          }
```

Similar thing for `scalePair` above.

bubblobill replied:
from the comment `/* limit scale to minimum 10%. Spinner changes are doubles, Slider changes are integers. */`, so a Number check won't work. This is actually checking the source of the event; if it is a Double it comes from a spinner, if it is an Integer it comes from a slider.
Since sliders and spinners update each other this will be fired multiple times so we only want to veto one of them.

kwvanderlinde replied:
:facepalm: Apparently I can't read.

kwvanderlinde reviewed (changes requested):
The code itself looks pretty much good from a couple minor things.

But there is some weirdness still going on with the layout panel related to the sliders and spinners:
1. Dragging any of the sliders do not cause the panel itself to update. I have to right-click or otherwise interact with the panel to get it to update. Sometimes I get torn rendering (partiial repaint) instead of a complete update.
2. If I drag the zoom slider all the way down, exceptions are shown in a dialog saying `width and height must be positive`. So presumably this value needs to be clamped somewhere above 0 before being used.
3. Similar thing with bringing the scale slider all the way down. An exception is logged (not shown to the user), and the image snaps back to full size.

cwisniew commented on src/main/java/net/rptools/maptool/client/ui/token/dialog/edit/TokenLayoutPanelHelper.java:
```diff
@@ -0,0 +1,1341 @@
+
+  private final String helpText = assembleHelpText();
+
+  /* Component Getters */
```
The type of comment that is not really needed as it provides no information about what the code is doing, and is likely to be inaccurate when other people edit the code / use IDE to create new getter

bubblobill replied:
Gotcha, remove road signs to confuse the enemy.

cwisniew commented on src/main/java/net/rptools/maptool/client/ui/token/dialog/edit/TokenLayoutPanelHelper.java:
```diff
@@ -0,0 +1,1341 @@
+    return scaleLabel;
+  }
+
+  /* Spinners */
```
The type of comment that is not really needed, and is likely to be inaccurate when other people edit the code / use IDE to create new getter, I will stop repeating this comment but it applies for all instances in this class

cwisniew commented on src/main/java/net/rptools/maptool/client/swing/SpinnerSliderPaired.java:
```diff
@@ -0,0 +1,555 @@
+      }
+    }
+
+    public NumericSliderModel sliderModelDelegate = new NumericSliderModel(0, 0, 0, 100);
```
The only public fields should be constants

cwisniew commented on src/main/java/net/rptools/maptool/client/swing/SpinnerSliderPaired.java:
```diff
@@ -0,0 +1,555 @@
+
+    // spotless:on
+    // @formatter:on
+    @Override
```
I can't see where spotless was turned off... In any case if it was it shouldn't have been :)

bubblobill replied:
See this tongue, see how it is all vibrating, surrounded by eructations, and accompanied by an displeasing sound? That is all for you.

cwisniew replied:
> See this tongue, see how it is all vibrating, surrounded by eructations, and accompanied by an displeasing sound? That is all for you.

Just let me know if I should be purchasing a perspex shield

cwisniew commented on src/main/java/net/rptools/maptool/client/ui/token/dialog/edit/EditTokenDialog.java:
```diff
@@ -211,7 +231,7 @@ public void closeDialog() {
     var combo = getStatSheetCombo();
     combo.removeAllItems();
-    // Default Entry
+    /* Default Entry */
```
um... why change from // to /* for single line comments? The same goes for all the instances below

bubblobill replied:
Because someone sent me on a mission to find commented out code, so I changed most single line comments to block comments to make it easier to search the code-base for `//`

cwisniew commented on src/main/java/net/rptools/maptool/client/ui/token/dialog/edit/TokenLayoutPanelHelper.java:
```diff
@@ -0,0 +1,1341 @@
+  }
+
+  /* used to determine which scale is being edited: 0 is XY, 1 is X, 2 is Y */
+  private int scaleAxis = 0;
```
This would be clearer as an enum

bubblobill replied:
This is inherited from ImageUtil.flip(BufferedImage image, int direction) which I renamed to flipCartesian.
Should the enum go there, or because it is fundamentally a property of a token's deisplay characteristics, should it live on token?
Because, you know, someone told me we can't change Token right now.

cwisniew replied:
> This is inherited from ImageUtil.flip(BufferedImage image, int direction) which I renamed to flipCartesian. Should the enum go there, or because it is fundamentally a property of a token's deisplay characteristics, should it live on token? Because, you know, someone told me we can't change Token right now.

I mean you have already made other changes to the Token class :)

bubblobill replied:
That was before you said stop. When you did I thought this PR was dead in the water.

kwvanderlinde replied:
It's fine to modify logic in `Token` and other model classes. As long as there aren't changes to the fields and we're not storing different concrete types in the fields, then saving will not be affected.

cwisniew replied:
> It's fine to modify logic in `Token` and other model classes. As long as there aren't changes to the fields and we're not storing different concrete types in the fields, then saving will not be affected.

There are new fields for facing, scale, and rotation, so anything modified will possibly not appear the same when loading into an old campaign, so this change is already excluded from 1.18, so it can't hurt to make the other change its already going to be delayed to 1.19

Unless these shouldn't be persisted, in which case we should make them transient

bubblobill replied:
Moved to AppConstants as an Enum

cwisniew commented on src/main/java/net/rptools/maptool/client/ui/token/dialog/edit/TokenLayoutPanelHelper.java:
```diff
@@ -0,0 +1,1341 @@
+        .addFocusListener(focusListener);
+  }
+
+  public void initButtons() {
```
Only make it public if you are calling it from other classes (and then classes outside of the package / inheritance hierarchy). Same comment for all the Other init*() methods.

cwisniew commented on src/main/java/net/rptools/maptool/client/ui/zone/renderer/TokenLocation.java:
```diff
@@ -20,7 +20,7 @@
 import net.rptools.maptool.model.Token;
 
-class TokenLocation {
+public class TokenLocation {
```
Does this need to be made public? (I may have missed where you are creating/using it outside of the package). Someone make it package scope intentionally

bubblobill replied:
The decomposition of ZoneRenderer is naturally going to lead to urban sprawl. It makes sense (to me) to package things further to keep related stuff together. 
```
zone -> renderer -> tokenRendering  -> TokenRenderer
                                                             -> FacingArrowRenderer
                                                             -> HaloRendering
                             -> lighting ->
                             -> stuff
```
If it isn't public, descendant packages can't read it. 
TokenRenderer uses a private instance for class use. With no public TokenLocation it gets nerfed. 
[Hence this](https://discord.com/channels/296230822262865920/494598206626463755/1318088889457573969)

cwisniew replied:
> Hence this

But it doesn't as they are two different things and would not use the same class if it were public

bubblobill replied:
Renamed to TokenPosition

cwisniew commented on src/main/java/net/rptools/maptool/util/GraphicsUtil.java:
```diff
@@ -449,4 +445,60 @@ protected void paintComponent(Graphics g) {
     f.setVisible(true);
   }
+
+  public static Shape createGridShape(String gridType, double size) {
```
This should all be delegated to the Grid classes

bubblobill replied:
I disagree. 

1. The shapes generated by the Grid classes are woefully inadequate for close-up work because they use integer coordinates. It was one of my gripes when working up a GridRenderer, Integers do not make for pretty lines.
2. Without an instance of the Grid, you cannot access the grid shape. Creating a Grid in order to draw it's cell shape in a panel is overkill. IIRC The footprint editor does this with an instance of every Grid type.
3. I would concede it perhaps being a static method in Grid if anything. This would allow grids to summon a shape instead of having to rewrite the cell geometry creation code for each class.

cwisniew replied:
> I disagree.
> 
> 1. The shapes generated by the Grid classes are woefully inadequate for close-up work because they use integer coordinates. It was one of my gripes when working up a GridRenderer, Integers do not make for pretty lines.

Who said anything about them having to be integers just because they are part of the grid class?

> 2. Without an instance of the Grid, you cannot access the grid shape. Creating a Grid in order to draw it's cell shape in a panel is overkill. IIRC The footprint editor does this with an instance of every Grid type.
> 3. I would concede it perhaps being a static method in Grid if anything. This would allow grids to summon a shape instead of having to rewrite the cell geometry creation code for each class.

If you don't want to put them in the grid classes you could I guess create other classes where there is one for every grid type and a factory creates them, but is that more work then in grid? Certianly we shouldn't be creating a utility class that has a single function to do this that needs to be modified any time grids are added or modified.

bubblobill replied:
Moved it to Grid.

cwisniew commented on src/main/java/net/rptools/maptool/client/swing/SpinnerSliderPaired.java:
```diff
@@ -0,0 +1,555 @@
+    }
+
+    // Delegate classes with additional setters and normal setters redirected
+    public NumericSpinnerModel spinnerModelDelegate = new NumericSpinnerModel(0d, 0d, 100d, 1d);
```
Non constant fields should not be public

cwisniew commented on src/main/java/net/rptools/maptool/client/ui/token/dialog/edit/TokenLayoutPanelHelper.java:
```diff
@@ -0,0 +1,1341 @@
+  private static final double DEFAULT_FONT_SIZE = UI_DEFAULTS.getFont("defaultFont").getSize2D();
+  private static final int ICON_SIZE = (int) (DEFAULT_FONT_SIZE * 2 + 4);
+
+  Grid grid = MapTool.getFrame().getCurrentZoneRenderer().getZone().getGrid();
+  RenderBits renderBits = new RenderBits();
+  boolean isoGrid = grid.isIsometric();
+  boolean noGrid = GridFactory.getGridType(grid).equals(GridFactory.NONE);
+  boolean squareGrid = GridFactory.getGridType(grid).equals(GridFactory.SQUARE);
+  boolean hexGrid = false;
+  boolean isIsoFigure = false;
+  int gridSize = grid.getSize();
+  double cellHeight = grid.getCellHeight();
+  double cellWidth = grid.getCellWidth();
```
Are the access levels here (and elsewhere in the class) correct? Did you mean for them to be available to all classes in the same package? or should they be private?

bubblobill replied:
Most of them were shared at some point as I slowly split out the helper functions from the panel class.
So not in the least bit correct any more.

bubblobill replied:
Fixed

cwisniew commented on src/main/java/net/rptools/maptool/model/Token.java:
```diff
@@ -2217,6 +2128,7 @@ public String toString() {
     return "Token: " + id;
   }
 
+  // Token Layout - Getters/Setters
```
Not useful and will become inaccurate quickly

cwisniew commented on src/main/java/net/rptools/maptool/model/Token.java:
```diff
@@ -2227,27 +2139,127 @@ public Point getAnchor() {
+   * @return null or angle in degrees
+   */
+  public int getFacing() {
+    // -90° is natural alignment. TODO This should really be a per grid setting
```
Please don't add TODO comments into the code, if its really a TODO add an issue

cwisniew commented on src/main/java/net/rptools/maptool/model/Token.java:
```diff
@@ -2269,8 +2281,8 @@ public TokenTransferData toTransferData() {
     td.put(TokenTransferData.SNAP_TO_SCALE, snapToScale);
-    td.put(TokenTransferData.WIDTH, scaleX);
-    td.put(TokenTransferData.HEIGHT, scaleY);
+    td.put(TokenTransferData.WIDTH, scaleX); // this makes no sense
```
You are correct; adding this comment makes no sense at all, If you think its a bug test with the correct value to see result or add an issue

cwisniew commented on src/main/java/net/rptools/maptool/model/Token.java:
```diff
@@ -2269,8 +2281,8 @@ public TokenTransferData toTransferData() {
-    td.put(TokenTransferData.WIDTH, scaleX);
-    td.put(TokenTransferData.HEIGHT, scaleY);
+    td.put(TokenTransferData.WIDTH, scaleX); // this makes no sense
+    td.put(TokenTransferData.HEIGHT, scaleY); // ditto
```
ditto :)

bubblobill committed:
- Finished rename to setLayoutProps function (tested)

bubblobill commented:
> The code itself looks pretty much good from a couple minor things.
> 
> But there is some weirdness still going on with the layout panel related to the sliders and spinners:
> 
> 1. Dragging any of the sliders do not cause the panel itself to update. 
I appear to have failed to implement a change listener on the slider.

... Sometimes I get torn rendering (partiial repaint) instead of a complete update.
Odd, repaint should be performed on the control and the display panel. Hopefully the shiny new change listener will cause the region to be marked dirty and fix this.

> 2. If I drag the zoom slider all the way down, exceptions are shown in a dialog saying `width and height must be positive`. So presumably this value needs to be clamped somewhere above 0 before being used.
Remember asking me why I had error handling on the image scaling op?

> 3. Similar thing with bringing the scale slider all the way down. An exception is logged (not shown to the user), and the image snaps back to full size.
The full size thing is me just returning the original image when catching the error. The vetoed change should be catching this before it goes to rescaling.

bubblobill committed:
changed access thingies

bubblobill committed:
added clamp to zoom calculation

bubblobill committed:
added isDouble

bubblobill committed:
Gave up on veto handling for stopping things reaching undesirable values
. Added floor and ceiling properties to the numeric model instead.
Major rework of SpinnerSliderPaired
Spotless and tweaks

bubblobill committed:
removed debug logging calls

bubblobill committed:
Refactor: createGridShape moved to Grid

bubblobill committed:
Spotless

bubblobill committed:
Refactor: rename TokenLocation to TokenPosition

bubblobill committed:
Refactor: made flip direction into an enum

bubblobill committed:
spot

bubblobill committed:
DTO fix missing proprty
